### PR TITLE
For #8540: Update tracking protection policy on options selection!

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by mcarare** [comment]
@Amejia481 The first issue is solved in https://github.com/mozilla-mobile/fenix/pull/8687
The second in: https://github.com/mozilla-mobile/fenix/pull/8573

File: app/src/main/java/org/mozilla/fenix/components/Core.kt [hunk 1]
```diff
@@ -278,13 +278,17 @@ class Core(private val context: Context) {
     }
 
     private fun geCustomCookiePolicy(): CookiePolicy {
-            return when (context.settings().blockCookiesSelectionInCustomTrackingProtection) {
+        return if (context.settings().blockCookiesInCustomTrackingProtection) {
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/components/Core.kt]: I think we need to improve the logic here as we only want to set `CookiePolicy.ACCEPT_ALL` when the **Custom -> Cookies checkbox** is unchecked this is going to return true every time the checkbox is checked, and by default `blockTrackingContentInCustomTrackingProtection` [is set to true](https://github.com/mozilla-mobile/fenix/blob/388c7017d8378500eb986e5473465572055d804a/app/src/main/java/org/mozilla/fenix/utils/Settings.kt#L254)

Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/components/Core.kt]: I think we need to improve the logic here as we only want to set `CookiePolicy.ACCEPT_ALL` when in the UI ETP the **Custom Radio Button** is marked and the **Cookies checkbox** is unchecked this is going to return true every time the checkbox is checked, and by default `blockCookiesInCustomTrackingProtection` [is set to true](https://github.com/mozilla-mobile/fenix/blob/388c7017d8378500eb986e5473465572055d804a/app/src/main/java/org/mozilla/fenix/utils/Settings.kt#L245)

Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/components/Core.kt]: I think we need to improve the logic here as we only want to set `CookiePolicy.ACCEPT_ALL` when in the  ETP UI the **Custom Radio Button** is marked and the **Cookies checkbox** is unchecked this is going to return true every time the checkbox is checked, and by default `blockCookiesInCustomTrackingProtection` [is set to true](https://github.com/mozilla-mobile/fenix/blob/388c7017d8378500eb986e5473465572055d804a/app/src/main/java/org/mozilla/fenix/utils/Settings.kt#L245)

File: app/src/main/java/org/mozilla/fenix/components/TrackingProtectionPolicyFactory.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+package org.mozilla.fenix.components
+
+import mozilla.components.concept.engine.EngineSession.TrackingProtectionPolicy
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * Handles the logic behind creating new [TrackingProtectionPolicy]s.
+ */
+class TrackingProtectionPolicyFactory(private val settings: Settings) {
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/components/TrackingProtectionPolicyFactory.kt]: Moved the logic into this class to make it easier to test.

File: app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt [hunk 3]
```diff
@@ -213,34 +214,66 @@ class TrackingProtectionFragment : PreferenceFragmentCompat() {
 
         customCookies.onPreferenceChangeListener = object : SharedPreferenceUpdater() {
             override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
-                    updateTrackingProtectionPolicy()
                 customCookiesSelect.isVisible = !customCookies.isChecked
-                return super.onPreferenceChange(preference, newValue)
+                return super.onPreferenceChange(preference, newValue).also {
+                    updateTrackingProtectionPolicy()
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt]: Since `updateTrackingProtectionPolicy` uses information from shared prefs, it's important that we call super (and write to SP) before calling it.

File: app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt [hunk 4]
```diff
@@ -213,34 +214,66 @@ class TrackingProtectionFragment : PreferenceFragmentCompat() {
...
+                }
             }
         }
 
         customCookiesSelect.onPreferenceChangeListener = object : StringSharedPreferenceUpdater() {
             override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
-                updateTrackingProtectionPolicy()
-                val newValueEntry = (preference as DropDownListPreference).findEntry(key = newValue)
-                return super.onPreferenceChange(preference, newValueEntry)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt]: This whole ETP thing is really complex and hard to follow.  I think we were mistaken before @mcarare, this is working fine without any of the preference key/value mapping.

**Comment by Baron-Severin** [comment]
@mcarare would you mind looking this over?  I've tested and https://github.com/mozilla-mobile/fenix/issues/8594, https://github.com/mozilla-mobile/fenix/issues/8540, and https://github.com/mozilla-mobile/fenix/issues/8565 all seem to be working properly, but it's been a long day and I could have missed something.

Be aware that social trackers won't be blocked in custom mode.  There's already a patch to fix it here https://github.com/mozilla-mobile/android-components/pull/6080/.

File: app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt (hunk 4 above)
Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt]: Yes, I think it handles that internally. The initial problem was using `SharedPreferenceUpdater` that only handles booleans, not string. With the new `StringSharedPreferenceUpdater` it works without the mapping.

File: app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt [hunk 5]
```diff
@@ -103,16 +103,18 @@ class TrackingProtectionFragment : PreferenceFragmentCompat() {
             getString(R.string.preference_enhanced_tracking_protection_strict_info_button)
         radioStrict.onPreferenceChangeListener = object : SharedPreferenceUpdater() {
             override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
-                if (newValue == true) {
-                    updateTrackingProtectionPolicy()
-                    context?.metrics?.track(
-                        Event.TrackingProtectionSettingChanged(
-                            Event.TrackingProtectionSettingChanged.Setting.STRICT
+                return super.onPreferenceChange(preference, newValue).also {
```
Comment (mcarare) [inline app/src/main/java/org/mozilla/fenix/settings/TrackingProtectionFragment.kt]: @Baron-Severin @Amejia481  I moved  ` updateTrackingProtectionPolicy()` after the preference change, like Severin recommended for the checkboxes. Also, I've debugged this and with previous setting all the other preference keys remained true even if visually the radioButton was unchecked. That is why I added the `setCheckedWithoutClickListener` call. Any thoughts?

File: app/src/main/java/org/mozilla/fenix/components/TrackingProtectionPolicyFactory.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    @Suppress("ComplexMethod")
+    fun createTrackingProtectionPolicy(
+        normalMode: Boolean = settings.shouldUseTrackingProtection,
+        privateMode: Boolean = true
+    ): TrackingProtectionPolicy {
+        val trackingProtectionPolicy =
+            when {
+                settings.useStrictTrackingProtection -> TrackingProtectionPolicy.strict()
+                settings.useCustomTrackingProtection -> return TrackingProtectionPolicy.select(
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/components/TrackingProtectionPolicyFactory.kt]: nit: Could we  extract the custom tracking protection policy into a new function

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt [hunk 7]
```diff
@@ -247,9 +247,10 @@ class Settings private constructor(
         appContext.getPreferenceKey(R.string.pref_key_tracking_protection_custom_cookies),
         true
     )
+
     val blockCookiesSelectionInCustomTrackingProtection by stringPreference(
         appContext.getPreferenceKey(R.string.pref_key_tracking_protection_custom_cookies_select),
-        ""
+        "all"
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: "all" is to aggressive to be the default we should have "social" -> "Cross-site and Social media trackers" as default following the same order as the spinner

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: Also I noticed in desktop if we switch policies like from **Custom** to **Strict** or **Standard** the custom values return to the default values, I think we need to include this functionality, as the user could get a lot of site breakage if we keep the previous values as the user could think that is our default values

Comment (mcarare) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: > Also I noticed in desktop if we switch policies like from **Custom** to **Strict** or **Standard** the custom values return to the default values, I think we need to include this functionality, as the user could get a lot of site breakage if we keep the previous values as the user could think that is our default values

We've discussed this and this behaviour remains TBD. Thanks @Amejia481 😄

Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: 👍

**Comment by ekager** [comment]
Looks like there are some extraneous commits here. Could we rebase and push before merging?

**Comment by Baron-Severin** [comment]
> Looks like there are some extraneous commits here. Could we rebase and push before merging?

Removed it then squashed.
